tensorflow/core/kernels/training_ops.cc

```diff
@@ -310,21 +310,63 @@ struct ApplyAdamNonCuda {
                   typename TTypes<T>::ConstScalar beta2,
                   typename TTypes<T>::ConstScalar epsilon,
                   typename TTypes<T>::ConstFlat grad, bool use_nesterov) {
+    const int length = var.size();
+
+    T* var_ptr = var.data();
+    T* m_ptr = m.data();
+    T* v_ptr = v.data();
+    const T* g_ptr = grad.data();
     const T alpha = lr() * Eigen::numext::sqrt(T(1) - beta2_power()) /
                     (T(1) - beta1_power());
     // beta1 == μ
     // beta2 == ν
     // v     == n
     // var   == θ
 
-    m.device(d) += (grad - m) * (T(1) - beta1());
-    v.device(d) += (grad.square() - v) * (T(1) - beta2());
-    if (use_nesterov) {
-      var.device(d) -= ((grad * (T(1) - beta1()) + beta1() * m) * alpha) /
-                       (v.sqrt() + epsilon());
-    } else {
-      var.device(d) -= (m * alpha) / (v.sqrt() + epsilon());
-    }
+    auto shard = [this, var_ptr, m_ptr, v_ptr, g_ptr, alpha, beta1, beta2,
+                  epsilon, use_nesterov](int begin, int end) {
+      T m;
+      T v;
+      T g;
+
+      if (use_nesterov) {
```

Comment:
I think both of these loops could be vectorized/simplified with Eigen (see example from redux_functor.h: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/redux_functor.h#L72-L95)

```
    using Input = Eigen::TensorMap<
        Eigen::Tensor<const T, 1, Eigen::RowMajor, Eigen::Index>,
        Eigen::Unaligned>;

    Input m(m_ptr, (begin-end)), v(..), g(..);

    if (use_nesterov) {
      m += (grad - m) * (T(1) - beta1());
      v += (grad.square() - v) * (T(1) - beta2());
      var -= ((grad * (T(1) - beta1()) + beta1() * m) * alpha) / (v.sqrt() + epsilon());
    } else {
       ...
    }
}
```

just remove `.device()` from tensor expressions to evaluate them with default single threaded device

Reply:
> I think both of these loops could be vectorized/simplified with Eigen (see example from redux_functor.h: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/redux_functor.h#L72-L95)
> 
> ```
>     using Input = Eigen::TensorMap<
>         Eigen::Tensor<const T, 1, Eigen::RowMajor, Eigen::Index>,
>         Eigen::Unaligned>;
> 
>     Input m(m_ptr, (begin-end)), v(..), g(..);
> 
>     if (use_nesterov) {
>       m += (grad - m) * (T(1) - beta1());
>       v += (grad.square() - v) * (T(1) - beta2());
>       var -= ((grad * (T(1) - beta1()) + beta1() * m) * alpha) / (v.sqrt() + epsilon());
>     } else {
>        ...
>     }
> }
> ```
> 
> just remove `.device()` from tensor expressions to evaluate them with default single threaded device

Good idea. I also find if use "device.parallelFor" instead "Shard", the function parameter can keep unchanged.

Reply:
I realized that memory access pattern might be not as good as in a manual loop, but I'd love to see proofs in micro benchmarks. If there is no performance penalty I'd prefer smaller/simpler to read code.

Reply:
OK, after I refine this code I'll try to make a micro benchmark, please wait 1~2 days, thanks.

Reply:
Done. Please check the new code, thanks!